packages/dds/merge-tree/src/revertibles.ts

```diff
@@ -0,0 +1,377 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { assert, unreachableCase } from "@fluidframework/common-utils";
+import { UsageError } from "@fluidframework/container-utils";
+import { List } from "./collections";
+import { EndOfTreeSegment } from "./endOfTreeSegment";
+import { LocalReferenceCollection, LocalReferencePosition } from "./localReference";
+import {
+    IMergeTreeDeltaCallbackArgs,
+} from "./mergeTreeDeltaCallback";
+import { IRemovalInfo, ISegment, toRemovalInfo } from "./mergeTreeNodes";
+import { depthFirstNodeWalk } from "./mergeTreeNodeWalk";
+import { TrackingGroup } from "./mergeTreeTracking";
+import {
+    IJSONSegment,
+    MergeTreeDeltaType,
+    ReferenceType,
+} from "./ops";
+import { matchProperties, PropertySet } from "./properties";
+import { DetachedReferencePosition } from "./referencePositions";
+
+/**
+ * @alpha Revertibles are new and require the option
+ *  mergeTreeUseNewLengthCalculations to be set as true on the underlying merge tree
+ *  in order to function correctly.
+ */
+export type MergeTreeDeltaRevertible =
+    {
+        operation: typeof MergeTreeDeltaType.INSERT;
+        trackingGroup: TrackingGroup;
+    } | {
+        operation: typeof MergeTreeDeltaType.REMOVE;
+        trackingGroup: TrackingGroup;
+    } | {
+        operation: typeof MergeTreeDeltaType.ANNOTATE;
+        trackingGroup: TrackingGroup;
+        propertyDeltas: PropertySet;
+    };
+
+type TypedRevertible<T extends MergeTreeDeltaRevertible["operation"]> =
+    MergeTreeDeltaRevertible & { operation: T; };
+
+interface RemoveSegmentRefProperties{
+    /**
+     * the serialized form of the segment, so it can be re-inserted
+     */
+    segSpec: IJSONSegment;
+    /**
+     * a tag  so the reference can be identified as being created for revert
+     */
+    referenceSpace: "mergeTreeDeltaRevertible";
+}
+
+/**
+ * @alpha Revertibles are new and require the option
+ *  mergeTreeUseNewLengthCalculations to be set as true on the underlying merge tree
+ *  in order to function correctly.
+ */
+export interface MergeTreeRevertibleDriver{
+    insertFromSpec(pos: number, spec: IJSONSegment);
+    removeRange(start: number, end: number);
+    annotateRange(
+        start: number,
+        end: number,
+        props: PropertySet);
+    createLocalReferencePosition(
+        segment: ISegment,
+        offset: number,
+        refType: ReferenceType,
+        properties: PropertySet | undefined): LocalReferencePosition;
+    localReferencePositionToPosition(lref: LocalReferencePosition): number;
+    getPosition(segment: ISegment): number;
+    getContainingSegment(pos: number): { segment: ISegment | undefined; offset: number | undefined; };
+}
+
+type InternalRevertDriver = MergeTreeRevertibleDriver & {
+    __mergeTreeRevertible?: {
+        detachedReferences?: EndOfTreeSegment;
```

Comment:
there are caveats with EndOfTreeSegment, but i did consider building it into merge tree, and using some other signal about which local references could go there. on of the major reasons i didn't was GC, as this could easily become a memory leak. once i move it here, i liked that is belongs to a specific domain, if we find the need for more usages, keeping them per domain is my preference, as it makes it such that we don't need to solve the a general gc problem here.